src/twister2/device/native_simulator_adapter.py

```diff
@@ -23,78 +25,86 @@ def __init__(self, twister_config, hardware_map: HardwareMap | None = None, **kw
-            raise
+        except subprocess.SubprocessError as e:
+            logger.error('Flashing failed due to subprocess error %s', e)
+            self._exc = TwisterFlashException(e.args)
+        except FileNotFoundError as e:
+            logger.error(f'Flashing failed due to file not found: {e.filename}')
+            self._exc = TwisterFlashException(f'File not found: {e.filename}')
+        except Exception as e:
+            logger.error('Flashing failed: %s', e)
+            self._exc = TwisterFlashException(e.args)
         else:
-            if self._process.returncode == 0:
-                logger.info('Finished flashing %s', build_dir)
-            elif self._process_ended_with_timeout:
-                raise TwisterTimeoutExpired(f'Process {self._process.pid} terminated after {timeout} seconds')
-            else:
-                if stderr := self._process.stderr.read():
-                    logger.error(stderr.decode())
-                raise TwisterFlashException(f'Could not run simulator with PID {self._process.pid}')
-
-    def _collect_process_output(self, process: subprocess.Popen) -> threading.Thread:
-        """Create Thread which saves a process output to a file."""
-        def _read():
-            with process.stdout:
-                for line in iter(process.stdout.readline, b''):
-                    self.queue.put(line.decode().strip())
-
-        return threading.Thread(target=_read, daemon=True)
+            if return_code == 0:
+                logger.info('Flashing finished with success')
+            elif return_code > 0:
+                self._exc = TwisterFlashException(f'Flashing finished with errors for PID {self._process.pid}')
+        finally:
+            self.queue.put(END_DATA)
+            self.queue.task_done()
+            self._process = None
 
-    def _wait_and_terminate_process(self, process: subprocess.Popen, timeout: float) -> threading.Thread:
-        """Create Thread which kills a process after given time."""
-        def _waiting():
-            end_time = time.time() + timeout
-            while process.poll() is None and time.time() < end_time:
-                time.sleep(0.1)
-            if process.poll() is None:
-                process.kill()
-                logger.error('Process %s terminated after %s seconds', process.pid, timeout)
-                self._process_ended_with_timeout = True
-        return threading.Thread(target=_waiting, daemon=True)
+    def disconnect(self):
+        """End subprocess."""
+        self.stop()
 
     @property
     def out(self) -> Generator[str, None, None]:
         """Return output from serial."""
-        while not self.queue.empty():
-            yield self.queue.get()
+        end_time = time.time() + self.timeout
+        while True:
+            line = self.queue.get()
+            if line == END_DATA:
+                break
+            yield line
+            if time.time() > end_time:
```

Comment:
Is this necessary to verify timeout here? It is monitored in `parse` method, so maybe it is redundant check. Otherwise we should have similar checking in `HardwareAdapter` `out` method.

Reply:
I think it can be remove from here.